GCS Set a default policy to retry failures

tigrux opened the pull request:
GCS: Add properties to configure retry policy

GCS may encounter recoverable errors, for example, an authentification error.
By default GCS keeps retrying for up to 15 minutes,
but this default may be excessive, giving the impression that Velox has become unresponsive.
This behaviour can be configured, with options to configure the time to keep retrying or the number of times to retry.
However, the GCS connector does not allow to configure neither the retry time nor the retry count.

This change introduces two new properties:
- hive.gcs.max-retry-count: integer
  The maximum retry counter of transient errors.
- hive.gcs.max-retry-time: integer
  The maximum time allowed (seconds) to retry transient errors.

Testing was performed manually, trying to read from an unauthorized bucket.

Fixes #9264

tigrux commented:
Hello @kgpai and @majetideepak.
We recently identified the GCS does not allow to configure the retry policy.
This PR adds a couple of properties to configure the retry policy.

kgpai commented:
cc: @majetideepak

majetideepak commented on velox/connectors/hive/HiveConfig.cpp:
```diff
@@ -117,6 +117,22 @@ std::string HiveConfig::gcsCredentials() const {
+}
+
+std::optional<int> HiveConfig::gcsMaxRetryTime() const {
+  folly::Optional<int> optional = config_->get<int>(kGCSMaxRetryTime);
```
same as above

tigrux replied:
More lines saved!

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:
```diff
@@ -269,6 +269,21 @@ class GCSFileSystem::Impl {
+    const auto max_retry_time = hiveConfig_->gcsMaxRetryTime();
+    if (max_retry_time) {
+      LOG(INFO) << "Config::gcsMaxRetryTime = " << max_retry_time.value();
+      const auto retry_time = std::chrono::seconds(max_retry_time.value());
```
no need of `const`

tigrux replied:
Fixed!

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/examples/GCSFileSystemExampleRead.cpp:
```diff
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include "velox/common/memory/Memory.h"
```
Why do we need this header?

tigrux replied:
Copy and paste is a destructive drug, don't consume drugs!

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/examples/GCSFileSystemExampleWrite.cpp:
```diff
@@ -0,0 +1,79 @@
+/*
```
Can we add this example in a separate PR? It does not relate to the new configs added here.

tigrux replied:
I wrote that example because it clearly shows the effect of leaving the default retry (15 minutes) and not authenticating, but you are right, I will provide it in another PR.

majetideepak reviewed (commented):
@tigrux some comments. Can we add a unit test for these configs?

tigrux commented:
> @tigrux some comments. Can we add a unit test for these configs?

I had to test manually, I do not know how to trigger a failure from the simulator.

tigrux commented:
@majetideepak @kgpai
I addressed most of the feedback, however, I was unable to add UTs (I could not get the simulator to trigger retries) so I added a note stating that I tested manually.

majetideepak commented on velox/docs/configs.rst:
```diff
@@ -547,6 +547,14 @@ Each query can override the config by setting corresponding query session proper
+     - integer
+     -
+     - The GCS maximum retry counter of transient errors.
+   * - hive.gcs.max-retry-time
```
We recently added a timeout for s3 using string. https://github.com/facebookincubator/velox/pull/9472
We could do the same here. Turns out it is straightforward.

tigrux replied:
Done!

majetideepak reviewed (commented):
@tigrux one comment. Thanks!

majetideepak commented on velox/connectors/hive/HiveConfig.cpp:
```diff
@@ -132,6 +132,14 @@ std::string HiveConfig::gcsCredentials() const {
+}
+
+std::optional<int> HiveConfig::gcsMaxRetryTime() const {
+  return static_cast<std::optional<int>>(config_->get<int>(kGCSMaxRetryTime));
```
Shouldn't this be a string?

tigrux replied:
Fixed!

majetideepak commented on velox/connectors/hive/HiveConfig.h:
```diff
@@ -230,6 +236,10 @@ class HiveConfig {
 
+  std::optional<int> gcsMaxRetryCount() const;
+
+  std::optional<int> gcsMaxRetryTime() const;
```
String?

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/examples/GCSFileSystemExample.cpp:
```diff
@@ -24,10 +24,19 @@
 
 DEFINE_string(gcs_path, "", "Path of GCS bucket");
+DEFINE_string(gcs_max_retry_count, "", "Max retry count");
+DEFINE_string(gcs_max_retry_time, "", "Max retry time (seconds)");
```
Can be any valid duration string. Let's remove seconds.

tigrux replied:
Fixed!

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:
```diff
@@ -275,6 +275,21 @@ class GCSFileSystem::Impl {
+    auto max_retry_time = hiveConfig_->gcsMaxRetryTime();
+    if (max_retry_time) {
+      LOG(INFO) << "Config::gcsMaxRetryTime = " << max_retry_time.value();
+      auto retry_time = std::chrono::seconds(max_retry_time.value());
```
Shouldn't this be
 ```
 std::chrono::duration_cast<std::chrono::milliseconds>(
              facebook::velox::core::toDuration())
``` 
?

majetideepak commented on velox/docs/configs.rst:
```diff
@@ -559,6 +559,14 @@ Each query can override the config by setting corresponding query session proper
      -
      - The GCS service account configuration as json string.
+   * - hive.gcs.max-retry-count
+     - string
```
This is still an integer right?

tigrux replied:
Fixed!

majetideepak replied:
I still see string here.

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:
```diff
@@ -275,6 +276,25 @@ class GCSFileSystem::Impl {
 
+    auto max_retry_count = hiveConfig_->gcsMaxRetryCount();
+    if (max_retry_count) {
+      LOG(INFO) << "Config::gcsMaxRetryCount = " << max_retry_count.value();
```
We don't log other config values (even for other FileSystems). Can we remove this logging here for now?
Same below.

tigrux replied:
Ouch! My mistake! I will fix it.

majetideepak reviewed (commented):
@tigrux 2 small comments.

majetideepak commented on velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:
```diff
@@ -275,6 +276,24 @@ class GCSFileSystem::Impl {
+    if (max_retry_time) {
+      auto retry_time = std::chrono::duration_cast<std::chrono::milliseconds>(
+          facebook::velox::core::toDuration(max_retry_time.value()));
+      const auto ms =
```
remove this log as well.

tigrux commented:
Hello @kgpai, could you take a look please? Thanks.
